File: 0141-linked-list-cycle/0141-linked-list-cycle.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */

/**
 * @param {ListNode} head
 * @return {boolean}
 */
var hasCycle = function(head) {
    let encountered_nodes = new Set([]);
    let current_node = head

    while (current_node) {
        if (encountered_nodes.has(current_node)){
            return true
        }
        encountered_nodes.add(current_node)
        current_node = current_node.next
    }

    return false

};

File: 0141-linked-list-cycle/0141-linked-list-cycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
        encountered_nodes = set()
        current_node = head
        while current_node:
            if current_node in encountered_nodes:
                return True
            
            encountered_nodes.add(current_node)
            current_node = current_node.next

        return False

        